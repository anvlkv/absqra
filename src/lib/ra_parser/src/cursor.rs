use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}