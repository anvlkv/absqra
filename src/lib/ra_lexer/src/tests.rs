use crate::cursor::{Cursor, Position};
use crate::{tokenize, Token, TokenKind};

mod cursor {
    use super::{Cursor, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 0, 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0, 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0, 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 0, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encountering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nSOME", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'S');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'O');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'M');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 0);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 0);
    }
}

mod lib {
    use super::{tokenize, Position, Token, TokenKind};
    

    #[test]
    fn it_should_create_iterator_of_tokens() {
        let mut stream = tokenize("abc");
        stream.next();
    }

    #[test]
    fn it_should_parse_multiple_lines() {
        let mut stream = tokenize("!?&/\n\tabc");
        stream.next();
        stream.next();
        stream.next();
        stream.next();
        assert_eq!(
            stream.next().unwrap(),
            Token {
                kind: Some(TokenKind::Identifier("abc")),
                content: "abc",
                position: (Position(2, 1), Position(2, 4)),
                len: 3,
                level: 1,
            }
        )
    }

    #[test]
    fn it_should_parse_identifiers() {
        let mut stream = tokenize("abc");
        assert_eq!(
            stream.next().unwrap(),
            Token {
                kind: Some(TokenKind::Identifier("abc")),
                content: "abc",
                position: (Position(1, 0), Position(1, 3)),
                len: 3,
                level: 0
            }
        );
    }

    use ra_common::insta::assert_debug_snapshot;
    use utils::for_each_ra_example_file;
    use std::fs::File;
    use std::io::Read;

    #[test]
    fn it_should_match_snapshots() {
        for_each_ra_example_file(|example| {
                let mut file = File::open(example.path()).unwrap();
    
                let mut contents = String::new();
                file.read_to_string(&mut contents).unwrap();

                println!("{}", example.file_name().to_str().unwrap());

                let tokens: Vec<Token> = tokenize(&contents).collect();
                assert_debug_snapshot!(example.path().to_str().unwrap(), tokens)
        });
    }

    mod strings {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_string_literals_with_double_quotes() {
            let mut stream = tokenize("\"some\"");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::StringLiteral("some")),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    ..Default::default()
                }
            )
        }

        #[test]
        fn it_should_parse_string_literals_with_single_quotes() {
            let mut stream = tokenize("\'some\'");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::StringLiteral("some")),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    ..Default::default()
                }
            )
        }
    }
    mod comments {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_single_line_comments() {
            let mut stream = tokenize("//abc");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 0
                }
            );
        }
        #[test]
        fn it_should_parse_multi_line_comments() {
            let mut stream = tokenize("/*abc\nSOME*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "abc\nSOME",
                    position: (Position(1, 0), Position(2, 6)),
                    len: 12,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_multi_line_comments_with_fringe() {
            let mut stream = tokenize("/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "\n* abc\n* SOME\n",
                    position: (Position(1, 0), Position(4, 2)),
                    len: 18,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_comments_followed_by_one_another() {
            let mut stream = tokenize("//some\n\n/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "some",
                    position: (Position(1, 0), Position(1, 6)),
                    len: 6,
                    level: 0
                }
            );

            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "\n* abc\n* SOME\n",
                    position: (Position(3, 0), Position(6, 2)),
                    len: 18,
                    level: 0
                }
            );
        }
    }
    
    mod numbers {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_numbers() {
            let mut stream = tokenize("123");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Int(123)),
                    content: "123",
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_separator() {
            let mut stream = tokenize("123,321");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123.321)),
                    content: "123,321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_separator() {
            let mut stream = tokenize("123.321");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123.321)),
                    content: "123.321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_thousands_separator() {
            let mut stream = tokenize("123.321,456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321.456)),
                    content: "123.321,456",
                    position: (Position(1, 0), Position(1, 11)),
                    len: 11,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_multiple_thousands_separator() {
            let mut stream = tokenize("123.321.123,456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321123.456)),
                    content: "123.321.123,456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_and_multiple_thousands_separators() {
            let mut stream = tokenize("123,321,123.456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321123.456)),
                    content: "123,321,123.456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 0
                }
            );
        }

        #[test]
        #[should_panic]
        fn it_should_panic_when_encountering_multiple_decimal_separators() {
            let mut stream = tokenize("123.321.123,456,654");
            stream.next();
        }

        #[test]
        fn it_should_parse_negative_integer() {
            let mut stream = tokenize("-123");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Int(-123)),
                    content: "-123",
                    position: (Position(1, 0), Position(1, 4)),
                    len: 4,
                    level: 0
                }
            )
        }

        #[test]
        fn it_should_parse_negative_float() {
            let mut stream = tokenize("-123.312");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(-123.312)),
                    content: "-123.312",
                    position: (Position(1, 0), Position(1, 8)),
                    len: 8,
                    level: 0
                }
            )
        }
    }

    mod content_block {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_content_blocks() {
            let mut stream = tokenize("`abc`");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::ContentBlock),
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_keep_inner_indents_when_parsing_content() {
            let mut stream = tokenize("`\n\tabc\n\t\tabc\n`");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::ContentBlock),
                    content: "\n\tabc\n\t\tabc\n",
                    position: (Position(1, 0), Position(4, 1)),
                    len: 14,
                    level: 0
                }
            );
        }

        #[test]
        #[should_panic]
        fn it_should_panic_when_content_block_is_not_closed_properly() {
            let mut stream = tokenize("`\n\tabc`");
            stream.next();
        }
    }

    mod single_char_tokens {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_tokens() {
            let mut stream = tokenize("!?&/#");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Exclamation),
                    position: (Position(1, 0), Position(1, 1)),
                    content: "!",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Question),
                    position: (Position(1, 1), Position(1, 2)),
                    content: "?",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Ampersand),
                    position: (Position(1, 2), Position(1, 3)),
                    content: "&",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Slash),
                    position: (Position(1, 3), Position(1, 4)),
                    content: "/",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::HashPound),
                    position: (Position(1, 4), Position(1, 5)),
                    content: "#",
                    level: 0,
                    ..Default::default()
                }
            );
        }
    }


}
