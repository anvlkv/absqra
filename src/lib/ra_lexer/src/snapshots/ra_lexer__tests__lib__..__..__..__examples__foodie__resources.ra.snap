---
source: src/lib/ra_lexer/src/tests.rs
expression: tokens
---
[
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        1,
        0
      ],
      [
        1,
        1
      ]
    ],
    "level": 0
  },
  {
    "kind": {
      "Identifier": "navigationItems"
    },
    "len": 15,
    "content": "navigationItems",
    "position": [
      [
        1,
        1
      ],
      [
        1,
        16
      ]
    ],
    "level": 0
  },
  {
    "kind": {
      "Identifier": "meals"
    },
    "len": 5,
    "content": "meals",
    "position": [
      [
        2,
        4
      ],
      [
        2,
        9
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        3,
        8
      ],
      [
        3,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        3,
        9
      ],
      [
        3,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        3,
        11
      ],
      [
        3,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        3,
        12
      ],
      [
        3,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 29,
    "content": "\n            Meals\n        ",
    "position": [
      [
        3,
        14
      ],
      [
        5,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        6,
        8
      ],
      [
        6,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        6,
        9
      ],
      [
        6,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        6,
        11
      ],
      [
        6,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        6,
        12
      ],
      [
        6,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 34,
    "content": "\n            Блюда\n        ",
    "position": [
      [
        6,
        14
      ],
      [
        8,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "countries"
    },
    "len": 9,
    "content": "countries",
    "position": [
      [
        9,
        4
      ],
      [
        9,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        10,
        8
      ],
      [
        10,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        10,
        9
      ],
      [
        10,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        10,
        11
      ],
      [
        10,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        10,
        12
      ],
      [
        10,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 33,
    "content": "\n            Countries\n        ",
    "position": [
      [
        10,
        14
      ],
      [
        12,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        13,
        8
      ],
      [
        13,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        13,
        9
      ],
      [
        13,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        13,
        11
      ],
      [
        13,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        13,
        12
      ],
      [
        13,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 59,
    "content": "\n            Национальные кухни\n        ",
    "position": [
      [
        13,
        14
      ],
      [
        15,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "restrictions"
    },
    "len": 12,
    "content": "restrictions",
    "position": [
      [
        16,
        4
      ],
      [
        16,
        16
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        17,
        8
      ],
      [
        17,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        17,
        9
      ],
      [
        17,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        17,
        11
      ],
      [
        17,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        17,
        12
      ],
      [
        17,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 35,
    "content": "\n            Preferences\n        ",
    "position": [
      [
        17,
        14
      ],
      [
        19,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        20,
        8
      ],
      [
        20,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        20,
        9
      ],
      [
        20,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        20,
        11
      ],
      [
        20,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        20,
        12
      ],
      [
        20,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 48,
    "content": "\n            Предпочтения\n        ",
    "position": [
      [
        20,
        14
      ],
      [
        22,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        24,
        0
      ],
      [
        24,
        1
      ]
    ],
    "level": 0
  },
  {
    "kind": {
      "Identifier": "userItems"
    },
    "len": 9,
    "content": "userItems",
    "position": [
      [
        24,
        1
      ],
      [
        24,
        10
      ]
    ],
    "level": 0
  },
  {
    "kind": {
      "Identifier": "addRecipe"
    },
    "len": 9,
    "content": "addRecipe",
    "position": [
      [
        25,
        4
      ],
      [
        25,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        26,
        8
      ],
      [
        26,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "kind"
    },
    "len": 4,
    "content": "kind",
    "position": [
      [
        26,
        9
      ],
      [
        26,
        13
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        26,
        13
      ],
      [
        26,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "image"
    },
    "len": 5,
    "content": "image",
    "position": [
      [
        26,
        14
      ],
      [
        26,
        19
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 44,
    "content": "\n            http://path.to.image\n        ",
    "position": [
      [
        26,
        19
      ],
      [
        28,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "Coma",
    "len": 1,
    "content": ",",
    "position": [
      [
        28,
        9
      ],
      [
        28,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        29,
        8
      ],
      [
        29,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        29,
        9
      ],
      [
        29,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        29,
        11
      ],
      [
        29,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        29,
        12
      ],
      [
        29,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 34,
    "content": "\n            Add recipe\n        ",
    "position": [
      [
        29,
        14
      ],
      [
        31,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        32,
        8
      ],
      [
        32,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        32,
        9
      ],
      [
        32,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        32,
        11
      ],
      [
        32,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        32,
        12
      ],
      [
        32,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 53,
    "content": "\n            Добавить рецепт\n        ",
    "position": [
      [
        32,
        14
      ],
      [
        34,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "myAccount"
    },
    "len": 9,
    "content": "myAccount",
    "position": [
      [
        35,
        4
      ],
      [
        35,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        36,
        8
      ],
      [
        36,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        36,
        9
      ],
      [
        36,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        36,
        11
      ],
      [
        36,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        36,
        12
      ],
      [
        36,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 34,
    "content": "\n            My account\n        ",
    "position": [
      [
        36,
        14
      ],
      [
        38,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        39,
        8
      ],
      [
        39,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        39,
        9
      ],
      [
        39,
        11
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        39,
        11
      ],
      [
        39,
        12
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        39,
        12
      ],
      [
        39,
        14
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 45,
    "content": "\n            Мой аккаунт\n        ",
    "position": [
      [
        39,
        14
      ],
      [
        41,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        43,
        0
      ],
      [
        43,
        1
      ]
    ],
    "level": 0
  },
  {
    "kind": {
      "Identifier": "searchPlaceholder"
    },
    "len": 17,
    "content": "searchPlaceholder",
    "position": [
      [
        43,
        1
      ],
      [
        43,
        18
      ]
    ],
    "level": 0
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        44,
        4
      ],
      [
        44,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        44,
        5
      ],
      [
        44,
        7
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        44,
        7
      ],
      [
        44,
        8
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        44,
        8
      ],
      [
        44,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 29,
    "content": "\n        Find a recipe\n    ",
    "position": [
      [
        44,
        10
      ],
      [
        46,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        47,
        4
      ],
      [
        47,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        47,
        5
      ],
      [
        47,
        7
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        47,
        7
      ],
      [
        47,
        8
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        47,
        8
      ],
      [
        47,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 39,
    "content": "\n        Найти рецепт\n    ",
    "position": [
      [
        47,
        10
      ],
      [
        49,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        51,
        0
      ],
      [
        51,
        1
      ]
    ],
    "level": 0
  },
  {
    "kind": {
      "Identifier": "pictureLabel"
    },
    "len": 12,
    "content": "pictureLabel",
    "position": [
      [
        51,
        1
      ],
      [
        51,
        13
      ]
    ],
    "level": 0
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        52,
        4
      ],
      [
        52,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        52,
        5
      ],
      [
        52,
        7
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        52,
        7
      ],
      [
        52,
        8
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        52,
        8
      ],
      [
        52,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 23,
    "content": "\n        Picture\n    ",
    "position": [
      [
        52,
        10
      ],
      [
        54,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        55,
        4
      ],
      [
        55,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        55,
        5
      ],
      [
        55,
        7
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        55,
        7
      ],
      [
        55,
        8
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        55,
        8
      ],
      [
        55,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 32,
    "content": "\n        Картинка\n    ",
    "position": [
      [
        55,
        10
      ],
      [
        57,
        5
      ]
    ],
    "level": 1
  }
]
